File: solution/053MaximumSubarray/053MaximumSubarray.go
package main

import "fmt"

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}
func maxSubArray(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	local := nums[0]
	global := nums[0]
	for i := 1; i < len(nums); i++ {
		//local optimal must contain nums[i]
		//so, if local<0, then only nums[i], otherwise nums[i]+local
		local = max(nums[i], nums[i]+local)
		global = max(global, local)
	}
	return global
}

func main() {
	fmt.Printf("%v\n", maxSubArray([]int{-2, 1, -3, 4, -1, 2, 1, -5, 4}))
}

File: solution/053MaximumSubarray/053MaximumSubarray.rs
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut local = nums[0];
        let mut global = nums[0];
        for i in 1..nums.len() {
            local = std::cmp::max(nums[i],local+nums[i]); 
            global = std::cmp::max(global,local);  
        }
        global
    }
}
